pandas/core/groupby/ops.py

```diff
@@ -844,18 +844,22 @@ def apply(
             if not mutated and not _is_indexed_like(res, group_axes, axis):
                 mutated = True
             result_values.append(res)
-
         # getattr pattern for __name__ is needed for functools.partial objects
         if len(group_keys) == 0 and getattr(f, "__name__", None) not in [
             "idxmin",
             "idxmax",
             "nanargmin",
             "nanargmax",
         ]:
-            # If group_keys is empty, then no function calls have been made,
-            #  so we will not have raised even if this is an invalid dtype.
-            #  So do one dummy call here to raise appropriate TypeError.
-            f(data.iloc[:0])
+            try:
+                #  If group_keys is empty, then no function calls have been made,
+                #  so we will not have raised even if this is an invalid dtype.
+                #  So do one dummy call here to raise appropriate TypeError.
+                f(data.iloc[:0])
```

Comment:
Should the `try/except IndexError` only be for `data.iloc[:0]`? A user's `f` could throw an `IndexError` which we don't necessarily want to capture here?

Reply:
Do we have a test that could demonstrate such an instance? I am having a little bit of an issue coming up with one.

Reply:
Maybe a UDF like:

```
def f(group):
   raise IndexError
```

Should still raise the `IndexError` and got be caught by this `try/except`?

Reply:
In version `1.3.5` before this PR #44092 was merged. The IndexError will not be raised (I just tested it out).

I believe the idea was that, regardless of the UD function, if the groupby has no groups then an empty dataframe is always returned.

Reply:
Okay. Nonetheless, if we know `data` is empty and `.iloc[:0]` will raise an `IndexError`, it might be more explicit to use

```
if not data.empty:
    f(data.iloc[:0])
```

I just want to avoid any unintended side effects of `except IndexError` catching the exception from `f` and not `data.iloc[:0]`

Reply:
Awesome idea. Will try it out.

Reply:
using `not data.empty` fails the tests for #44092.

Reply:
Which one in particular?

Reply:
I don't believe `data.iloc[:0]` should ever raise on any frame; is there an example where it does?

I do think we need to decide a general pattern of behavior for pandas methods that take a UDF and operate on an empty frame. I'm planning on putting a proposal together but it will be some time (maybe a week or two).

It seems to me catching some error types like IndexError but not others might be confusing for users. I believe this code was added in #44092 (cc @jbrockmendel) to handle cases where `apply` was called with a pandas method (e.g. "skew"). Can we instead add this block to just those particular methods?

Reply:
> Can we instead add this block to just those particular methods?

Works for me.